qiskit_ode/signals/signals.py

```diff
@@ -327,202 +447,693 @@ def add_samples(self, start_sample: int, samples: List):
-        signal: The type will depend on the given base class.
+    def envelope(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Return an array of the envelope values of each component."""
+        return np.moveaxis(self._eval_envelopes(t), 0, -1)
 
-    Raises:
-        Exception: if signals cannot be added.
-    """
+    def complex_value(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Return the sum of the complex values of each component."""
+        if dispatch.default_backend() == "jax":
+            t = Array(t)
+        exp_phases = np.exp(np.expand_dims(t, -1) * self._carrier_arg + self._phase_arg)
+        return np.sum(self.envelope(t) * exp_phases, axis=-1)
 
-    # ensure both arguments are signals
-    if isinstance(sig1, (int, float, complex)):
-        sig1 = Constant(sig1)
-    if isinstance(sig2, (int, float, complex)):
-        sig2 = Constant(sig2)
+    def __str__(self):
+        if self.name is not None:
+            return str(self.name)
 
-    return Signal(
-        lambda t: sig1.value(t) + sig2.value(t),
-        0,
-        0,
-    )
+        if len(self) == 0:
+            return "SignalSum()"
 
+        default_str = str(self[0])
+        for sig in self.components[1:]:
+            default_str += " + {}".format(str(sig))
 
-class VectorSignal:
-    """The vector version of the Signal class - the envelope is an array-valued
-    function, and carrier_freqs is an array of carrier frequencies.
+        return default_str
 
-    In addition, a drift_array is set to correspond to the value of the
-    VectorSignal when all "time-dependent terms" are off. E.g. if it is
-    composed of a list of Signal objects, this corresponds to the output when
-    all non-Constant signal objects are zero.
+    def flatten(self) -> Signal:
+        r"""Merge into a single ``Signal``\. The output frequency is given by the
+        average.
+        """
+
+        if len(self) == 0:
+            return Constant(0.0)
+        elif len(self) == 1:
+            return self.components[0]
+
+        ave_freq = np.sum(self.carrier_freq) / len(self)
+        shifted_arg = self._carrier_arg - (1j * 2 * np.pi * ave_freq)
+
+        def merged_env(t):
+            exp_phases = np.exp(np.expand_dims(Array(t), -1) * shifted_arg + self._phase_arg)
+            return np.sum(self.envelope(t) * exp_phases, axis=-1)
+
+        return Signal(envelope=merged_env, carrier_freq=ave_freq, name=str(self))
+
+
+class DiscreteSignalSum(DiscreteSignal, SignalSum):
+    """Represents a sum of discretized signals, all with the same
+    time parameters: dt, number of samples, and start time.
     """
 
     def __init__(
         self,
-        envelope: Callable,
-        carrier_freqs: Array,
-        phases: Array,
-        drift_array: Optional[Array] = None,
+        dt: float,
+        samples: Union[List, Array],
+        start_time: float = 0.0,
+        duration: int = None,
```

Comment:
It does not seem like this class is using duration. Is this needed here?

Reply:
Yeah it isn't ever used. It only serves as a possible convenience argument for specifying a `DiscreteSignal` of a given length where all samples are `0`. I'll remove it for now and we can see if we ever end up actually needing that.